shared/src/main/scala/mlscript/TypeDefs.scala

```diff
@@ -541,5 +542,180 @@ class TypeDefs extends ConstraintSolver { self: Typer =>
+      *   both if invariant position visit
+      */
+    def updateVariance(ty: SimpleType, curVariance: VarianceInfo)(implicit tyDef: TypeDef, visited: MutSet[Bool -> TypeVariable]): Unit = {
+      // helper method for calling updateVariance on fields
+      def fieldVarianceHelper(fieldTy: FieldType): Unit = {
+          fieldTy.lb.foreach(lb => updateVariance(lb, curVariance.flip))
+          updateVariance(fieldTy.ub, curVariance)
+      }
+
+      trace(s"updateVariance(ty: $ty of ${ty.getClass()}, curVariance: $curVariance)(${tyDef.nme.name}.${tyDef.tvarVariances})") {
+        ty match {
+          case ProxyType(underlying) => updateVariance(underlying, curVariance)
+          case TraitTag(_) | ClassTag(_, _) => ()
+          case ExtrType(pol) => ()
+          case t: TypeVariable =>
+            // update the variance information for the type variable
+            val oldVariance = tyDef.tvarVariances.getOrElseUpdate(t, VarianceInfo.bi)
+            val newVariance = oldVariance && curVariance
+            if (newVariance =/= oldVariance) {
+              tyDef.tvarVariances(t) = newVariance
+              varianceUpdated = true
+            }
+            // check if a type variable has not been visited before with the current variance
+            val visitedBefore = curVariance match {
+                case (VarianceInfo(true, true)) => false
+                case (VarianceInfo(true, false)) => visited.contains(true -> t)
+                case (VarianceInfo(false, true)) => visited.contains(false -> t)
+                case (VarianceInfo(false, false)) => 
+                  visited.contains(true -> t) && visited.contains(false -> t)
+            }
+            if (!visitedBefore) {
+              // updated visited set
+              curVariance match {
+                case (VarianceInfo(true, true)) => ()
+                case (VarianceInfo(true, false)) => visited += true -> t
+                case (VarianceInfo(false, true)) => visited += false -> t
+                case (VarianceInfo(false, false)) => 
+                  visited += true -> t
+                  visited += false -> t
+              }
+              // handle cases where there is no recursive
+              // definition by recomputing udpateVariance for
+              // the type variable but this time it's been added
+              // to the visited set
+              curVariance match {
+                case VarianceInfo(true, true) => ()
+                case VarianceInfo(false, true) =>
+                    t.upperBounds.foreach(ub => updateVariance(ub, curVariance))
+                case VarianceInfo(true, false) =>
+                    t.lowerBounds.foreach(lb => updateVariance(lb, curVariance))
+                case VarianceInfo(false, false) =>
+                    t.upperBounds.foreach(ub => updateVariance(ub, curVariance))
+                    t.lowerBounds.foreach(lb => updateVariance(lb, curVariance))
+              }
+            }
```

Comment:
Type variable lower bounds have positive polarity and their upper bounds have negative polarity. So you should recurse in them accordingly (_not_ passing `curVariance`).

Also, all this code can be rewritten more concisely as follows:
```suggestion
            val (visitLB, visitUB) = (
              !curVariance.isContravariant && !visited(true -> t),
              !curVariance.isCovariant && !visited(false -> t),
            )
            if (visitLB) visited += true -> t
            if (visitUB) visited += false -> t
            if (visitLB) t.lowerBounds.foreach(lb => updateVariance(lb, VarianceInfo.co))
            if (visitUB) t.upperBounds.foreach(ub => updateVariance(ub, VarianceInfo.contra))
```